server/vcr-server/agent_webhooks/views.py

```diff
@@ -464,3 +408,193 @@ def handle_register_issuer(message):
+            post_method=True,
+            payload={"explain_ltxt": str(e)},
+            headers=settings.ADMIN_REQUEST_HEADERS,
+        )
+        resp.raise_for_status()
+        return Response({"success": False, "error": str(e)})
+
+    return Response(response_data)
+
+
+def receive_credential(cred_ex_id, cred_data, v=None):
+    existing = False
+    if PROCESS_INBOUND_CREDENTIALS:
+        credential = Credential(cred_data)
+
+        # sanity check that we haven't received this credential yet
+        cred_id = credential.thread_id
+        existing_credential = CredentialModel.objects.filter(credential_id=cred_id)
+        if 0 < len(existing_credential):
+            # TODO - credential already exists in the database, what to do?
+            LOGGER.error(" >>> Received duplicate for credential_id: "
+                         + cred_id + ", exch id: " + cred_ex_id)
+            existing = True
+            ret_cred_id = cred_id
+        else:
+            # new credential, populate database
+            credential = credential_manager.process(credential)
+            ret_cred_id = credential.credential_id
+    else:
+        ret_cred_id = cred_data["thread_id"]
+
+    # check if the credential is in the wallet already
+    if existing:
+        resp = call_agent_with_retry(
+            f"{settings.AGENT_ADMIN_URL}/credential/{ret_cred_id}",
+            post_method=False,
+            headers=settings.ADMIN_REQUEST_HEADERS,
+        )
+        if resp.status_code == 404:
+            existing = False
+
+    # Instruct the agent to store the credential in wallet
+    if not existing:
+        # post with retry - if returned status is 503 unavailable retry a few times
+        resp = call_agent_with_retry(
+            f"{settings.AGENT_ADMIN_URL}/issue-credential{'-' + v if v else ''}/records/{cred_ex_id}/store",
+            post_method=True,
+            payload={"credential_id": ret_cred_id},
+            headers=settings.ADMIN_REQUEST_HEADERS,
+        )
+        if resp.status_code == 404:
+            # TODO assume the credential exchange has completed?
+            resp = call_agent_with_retry(
+                f"{settings.AGENT_ADMIN_URL}/credential/{ret_cred_id}",
+                post_method=False,
+                headers=settings.ADMIN_REQUEST_HEADERS,
+            )
+            if resp.status_code == 404:
+                LOGGER.error(
+                    " >>> Error cred exchange id is missing but credential is not available for " + cred_ex_id + ", " + ret_cred_id)
+                return Response("Error cred exchange id is missing but credential is not available", status=status.HTTP_400_BAD_REQUEST)
+            pass
+        else:
+            resp.raise_for_status()
+
+    # You can include this exception to test error reporting
+    if RANDOM_ERRORS and 1 == random.randint(1, 50):
+        return random_exception(cred_ex_id, 'recieve_credential')
```

Comment:
Nitpick: could we incorporate the `1 == random.randint(1, 50)` into the utility functions?
That way at runtime that part will not be executed unless the random errors are turned on - I suspect it could have a significant impact when mass-issuing credentials.

So the call would become:
```python
if RANDOM_ERRORS:
   return random_exception(cred_ex_id, 'recieve_credential')
```
and the function definition would be something like:
```
def random_exception(cred_ex_id, method=""):
   if 1 == random.randint(1, 50):
      ....
```

Reply:
We could, one of the functions actually returns a response, so I'll do a second take on this